File: sdk/js/src/agent.ts
/**
 * `Agent` — mirrors `new Agent({ name, instructions, model, tools, handoffs })`
 * from the OpenAI Agents SDK.
 */

import type { AgentConfig, ClientOptions, MemoryConfig, ToolSpec } from "./types.js";

export class Agent {
  readonly name: string;
  readonly instructions?: string;
  readonly model?: string;
  readonly tools: ToolSpec[];
  readonly handoffs: Agent[];
  readonly handoffDescription?: string;
  readonly id?: string;
  readonly client: ClientOptions;
  /** Memory backend configuration (`cloud` / `local` / `both`). */
  readonly memory: MemoryConfig;

  constructor(config: AgentConfig) {
    if (!config.name || !config.name.trim()) {
      throw new Error("Agent requires a `name`");
    }
    this.name = config.name;
    this.instructions = config.instructions;
    this.model = config.model;
    this.tools = config.tools ?? [];
    this.handoffs = config.handoffs ?? [];
    this.handoffDescription = config.handoffDescription;
    this.id = config.id;
    this.client = config.client ?? {};
    this.memory = config.memory ?? {};
  }

  /** Static factory, matching `Agent.create({...})` in the Agents SDK. */
  static create(config: AgentConfig): Agent {
    return new Agent(config);
  }

  /** Clone with overrides (used when a handoff takes over a run). */
  clone(overrides: Partial<AgentConfig> = {}): Agent {
    return new Agent({
      name: overrides.name ?? this.name,
      instructions: overrides.instructions ?? this.instructions,
      model: overrides.model ?? this.model,
      tools: overrides.tools ?? this.tools,
      handoffs: overrides.handoffs ?? this.handoffs,
      handoffDescription: overrides.handoffDescription ?? this.handoffDescription,
      id: overrides.id ?? this.id,
      client: overrides.client ?? this.client,
      memory: overrides.memory ?? this.memory,
    });
  }

  /** JSON-Schema tool declarations forwarded to the cloud agent. */
  toolSchemas(): ToolSpec[] {
    return this.tools.map((t) => ({
      name: t.name,
      description: t.description,
      parameters: t.parameters ?? { type: "object", properties: {} },
    }));
  }
}

File: sdk/python/ariacompute_agent/agent.py
"""``Agent`` — mirrors ``Agent(...)`` from the OpenAI Agents SDK."""

from __future__ import annotations

from typing import Any, Optional, Sequence, Union

from .memory import MemoryBackend
from .tool import tool_from_function
from .types import ClientOptions, ToolSpec


class Agent:
    """A reusable agent definition.

    ```python
    agent = Agent(
        name="History tutor",
        instructions="Answer history questions clearly and concisely.",
        model="gpt-4o-mini",
    )
    ```
    """

    def __init__(
        self,
        name: str,
        instructions: Optional[str] = None,
        model: Optional[str] = None,
        tools: Optional[Sequence[Any]] = None,
        handoffs: Optional[Sequence["Agent"]] = None,
        handoff_description: Optional[str] = None,
        id: Optional[str] = None,
        client: Optional[ClientOptions] = None,
        memory_backend: Union[str, MemoryBackend, None] = None,
        memo_db: Optional[str] = None,
    ) -> None:
        if not name or not name.strip():
            raise ValueError("Agent requires a `name`")
        self.name = name
        self.instructions = instructions
        self.model = model
        self.tools = [self._coerce_tool(t) for t in (tools or [])]
        self.handoffs = list(handoffs or [])
        self.handoff_description = handoff_description
        self.id = id
        self.client = client or ClientOptions()
        self.memory_backend = (
            MemoryBackend.parse(memory_backend) if memory_backend is not None
            else MemoryBackend.CLOUD
        )
        self.memo_db = memo_db

    @staticmethod
    def _coerce_tool(t: Any) -> ToolSpec:
        if isinstance(t, ToolSpec):
            return t
        if callable(t):
            return tool_from_function(t)
        raise TypeError(f"unsupported tool: {t!r}")

    def clone(self, **overrides: Any) -> "Agent":
        """Return a copy with overridden fields (used by handoffs)."""
        params: dict[str, Any] = dict(
            name=self.name,
            instructions=self.instructions,
            model=self.model,
            tools=self.tools,
            handoffs=self.handoffs,
            handoff_description=self.handoff_description,
            id=self.id,
            client=self.client,
            memory_backend=self.memory_backend,
            memo_db=self.memo_db,
        )
        params.update(overrides)
        return Agent(**params)

    def tool_schemas(self) -> list[dict[str, Any]]:
        """JSON-schema tool declarations forwarded to the cloud."""
        return [t.schema() for t in self.tools]

    def __repr__(self) -> str:  # pragma: no cover - debugging aid
        return f"Agent(name={self.name!r}, model={self.model!r})"
